src/measure_theory/measurable_space.lean

```diff
@@ -352,6 +352,38 @@ lemma measurable_unit [measurable_space α] (f : unit → α) : measurable f :=
 have f = (λu, f ()) := funext $ assume ⟨⟩, rfl,
 by rw this; exact measurable_const
 
+section nat
+
+lemma measurable_from_nat [measurable_space α] {f : ℕ → α} : measurable f :=
+assume s hs, show is_measurable {n : ℕ | f n ∈ s}, from trivial
+
+lemma measurable_to_nat [measurable_space α] {f : α → ℕ} :
+(∀ k, is_measurable {x | f x = k }) → measurable f :=
+begin
+  assume h s hs, show is_measurable {x | f x ∈ s},
+  have : {x | f x ∈ s} = ⋃ (n ∈ s), {x | f x = n}, { ext, simp },
+  rw this, simp [is_measurable.Union, is_measurable.Union_Prop, h]
+end
+
+lemma measurable_find_greatest [measurable_space α] {p : ℕ → α → Prop} :
+  ∀ {N}, (∀ k ≤ N, is_measurable {x | nat.find_greatest (λ n, p n x) N = k}) →
```

Comment:
`∀k ≤ N, is_measurable {x | p k x}` should be enough as assumption.

Reply:
When proving the density of simple functions in borel measurable functions, I followed your proof in Isabelle and defined something similar to
``` isabelle
define A where [abs_def]: "A m n =
  {x∈space M. dist (f x) (e n) < 1 / (Suc m) ∧ 1 / (Suc m) ≤ dist (f x) z}" for m n
define B where [abs_def]: "B m = disjointed (A m)" for m

define m where [abs_def]: "m N x = Max {m. m ≤ N ∧ x ∈ (⋃n≤N. B m n)}" for N x
define F where [abs_def]: "F N x =
  (if (∃m≤N. x ∈ (⋃n≤N. B m n)) ∧ (∃n≤N. x ∈ B (m N x) n)
   then e (LEAST n. x ∈ B (m N x) n) else z)" for N x
```

I'm really impressed that in Isabelle, you proved the measurability of `F` in just one line:
``` isabelle
have [measurable]: "⋀i. F i ∈ borel_measurable M"
  unfolding F_def m_def by measurable
```
whereas I used more than a hundred lines. That is truly amazing.

Anyways, I don't like my formulation of `measurable_find_greatest` either. It looks awkward. And I know that `∀k ≤ N, is_measurable {x | p k x}` is enough. But too bad it is not enough to prove the measurability of `F`.

Reply:
I think a lot of the calculations you do are rolled into the supporting lemmas used by Isabelle's `measurable` tactic. It would be great to have a similar tactic for Lean. (@rwbarton's `continuity` tactic might serve as a template.) When this PR is settled, it would be good to look into that.

Reply:
A hundred lines still sounds excessive, if it can be one-shotted by a tactic. I would presume that the `measurable` tactic is just a compositional prover, so it should come out to two or three lines at most. So there must be something new lemma-wise that hasn't been noticed in the isabelle version.

Reply:
Yes, for all the occurring constants exist `measurable` rules.

Reply:
So Joe should really derive these rules and improve the proof. But this PR is already pretty unwieldy, so I think it is best to fix all the other issues, get it merged, and then follow up with another PR with the rules and a better proof.